crates/agent/src/controllers/activation.rs

```diff
@@ -115,23 +193,263 @@ pub async fn update_activation<C: ControlPlane>(
+        .as_ref()
+        .map(|check| check.last_ts)
+        .or(status.last_activated_at)
+    else {
+        anyhow::bail!(
+            "internal controller error: attempted to health check a spec that was never activated"
+        );
+    };
+
+    let next = from + wait_interval;
+    let now = control_plane.current_time();
+    if now < next {
+        // We're still waiting for the next health check
+        return Ok(Some(NextRun::after(next)));
+    }
+
+    // Check the status of the shards
+    let Some(task_type) = state
+        .live_spec
+        .as_ref()
+        .and_then(|s| to_ops_task_type(s.catalog_type()))
+    else {
+        anyhow::bail!("internal controller error: attempted to health check a catalog test");
+    };
+
+    let list_response = control_plane
+        .list_task_shards(state.data_plane_id, task_type, state.catalog_name.clone())
+        .await
+        .context("listing task shards")?;
+
+    if list_response.status != proto_gazette::consumer::Status::Ok as i32 {
+        return Err(anyhow::anyhow!(
+            "shard list response status not Ok, was {}",
+            list_response.status().as_str_name()
+        ))
+        .with_retry(NextRun::after_minutes(1))
+        .map_err(Into::into);
+    }
+
+    // Determine aggregate status from list response
+    let new_status = aggregate_shard_status(state.last_build_id, &list_response.shards);
+
+    let (first_ts, count) = match &status.shard_status {
+        Some(prev) if prev.status == new_status => (prev.first_ts, prev.count + 1),
+        Some(prev_status) => {
+            tracing::info!(?prev_status, prev_count = %prev_status.count, new_status = ?new_status, "task shard health status changed");
+            (now, 1)
+        }
+        None => (now, 1),
+    };
+    status.shard_status = Some(ShardStatusCheck {
+        count,
+        first_ts,
+        last_ts: now,
+        status: new_status,
+    });
+
+    // If there's been at least 3 failed checks in a row, and it's been longer than the
+    // re-activation interval, then re-activate failed shards. We require at least two
+    // failed checks in a row because sometimes shards can temporarily be in failed state
+    // just due normal infrastructure changes in the data plane.
+    let time_since_activation = now - status.last_activated_at.unwrap();
+    if new_status == ShardsStatus::Failed
+        && time_since_activation >= get_reactivate_interval(state)
+        && count >= 3
+    {
+        // If we've reached this section, then we have _not_ received any
+        // ShardFailed events, which would have caused us to set a `next_retry`.
+        // This can happen if the ops catalog is broken, or in certain edge
+        // cases in the data plane.
+        tracing::warn!(%time_since_activation, failed_checks = %count, "re-activating task shards because they still show as Failed after prior activation");
+        do_activate(now, state, status, control_plane).await?;
```

Comment:
discussion: a little hand-wavy, but you could have it send itself a ShardFailed event which would keep a single control flow for how the controller "observes" a failure event. Not asserting that's worth it / makes sense but it's an intuitive recursive framing to me at least. 

The flow in here is complex (but so is the domain), so i'm just casting about for any patterns which might help simplify.

Reply:
I did consider that, but only until it started getting a bit murky. I was uncertain about how to map an observation of Failed shards via listing into `ShardFailed` events, and saw a couple of challenges:

What if the actual `ShardFailed` event just gets materialized after we've observed the failed shards via listing? A slowdown in the ops catalog could result in duplicate `ShardFailed` events, which would result in longer restart backoffs. If there was a good way to deduplicate those ShardFailed events, we'd be all set, but I can't think of one. Definitely open to suggestions here, because I'm with you on wanting to reign in the complexity if we can.

Reply:
> What if the actual ShardFailed event just gets materialized after we've observed the failed shards via listing? A slowdown in the ops catalog could result in duplicate ShardFailed events, which would result in longer restart backoffs.

I don't think this is unique to this question -- for example, a reactor deployment will produce extra ShardFailed events just due to the elevated hand-offs.

Reply:
Then I'll take a crack at it, and see how it goes. Will update here once I try it out